server/core/src/main/java/org/infinispan/server/iteration/HashMapIterationManager.java

```diff
@@ -0,0 +1,76 @@
+package org.infinispan.server.iteration;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import org.infinispan.AdvancedCache;
+import org.infinispan.CacheStream;
+import org.infinispan.commons.time.TimeService;
+import org.infinispan.commons.util.IntSet;
+import org.infinispan.container.entries.CacheEntry;
+import org.infinispan.container.entries.ImmortalCacheEntry;
+import org.infinispan.security.actions.SecurityActions;
+import org.infinispan.stream.impl.local.AbstractLocalCacheStream;
+import org.infinispan.stream.impl.local.LocalCacheStream;
+
+public class HashMapIterationManager extends DefaultIterationManager {
+
+   private static final CacheStream<CacheEntry<Object, Object>> EMPTY_STREAM =
+         new LocalCacheStream<>(new HashMapStreamSupplier(null), false, null);
+
+   public HashMapIterationManager(TimeService timeService) {
+      super(timeService);
+   }
+
+   @Override
+   protected CacheStream<CacheEntry<Object, Object>> baseStream(AdvancedCache cache, IterationInitializationContext ctx) {
+      if (ctx == null) return EMPTY_STREAM;
+      if (!(ctx instanceof HashMapInitializationContext)) throw new IllegalStateException("Illegal context for iteration");
+
+      Map<Object, Object> source = ((HashMapInitializationContext) ctx).source;
+      AbstractLocalCacheStream.StreamSupplier<CacheEntry<Object, Object>, Stream<CacheEntry<Object, Object>>> ss = new HashMapStreamSupplier(source);
+      return new LocalCacheStream<>(ss, false, SecurityActions.getCacheComponentRegistry(cache));
```

Comment:
I'm wondering if we really need a full cache stream even to iterate on an inner struct

Reply:
This one also was a small paint point. I think this one might be necessary as we can use the existing code to apply the filters. But let me think about this, too.